x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/StreamingHttpResultPublisher.java

```diff
@@ -178,57 +128,161 @@ public boolean cancel() {
+                    }
+                }
+
+                @Override
+                public void cancel() {
+                    if (subscriptionCanceled.compareAndSet(false, true)) {
+                        taskRunner.cancel();
+                    }
+                }
+            });
+        }
+
+        @Override
+        public void onNext(byte[] item) {
+            contentQueue.offer(item);
+            taskRunner.requestNextRun();
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            if (throwable instanceof Exception e) {
+                pendingError = e;
             } else {
-                // per Subscription's spec, fail the subscriber and stop the processor
-                cancel();
-                subscriber.onError(new IllegalArgumentException("Subscriber requested a non-positive number " + n));
+                ExceptionsHelper.maybeError(throwable).ifPresent(ExceptionsHelper::maybeDieOnAnotherThread);
+                pendingError = new RuntimeException("Unhandled error while streaming");
             }
+            taskRunner.requestNextRun();
         }
 
         @Override
-        public void cancel() {
-            if (subscriptionCanceled.compareAndSet(false, true)) {
-                taskRunner.cancel();
-            }
+        public void onComplete() {
+            completed = true;
+            taskRunner.requestNextRun();
         }
-    }
 
-    private class OffloadThread implements Runnable {
         @Override
-        public void run() {
-            if (subscriptionCanceled.get()) {
-                return;
+        public void onSubscribe(Flow.Subscription subscription) {
+            assert false : "Apache never calls this.";
+            throw new UnsupportedOperationException("Apache never calls this.");
+        }
+    }
+
+    /**
+     * We want to keep track of how much memory we are consuming while reading the payload from the external provider. Apache continuously
+     * pushes payload data to us, whereas the client only requests the next set of bytes when they are ready, so we want to track how much
+     * data we are holding in memory and potentially pause the Apache client if we have reached our limit.
+     */
+    private static class ApacheClientBackpressure {
+        private final HttpSettings settings;
+        private final AtomicLong bytesInQueue = new AtomicLong(0);
+        private final Object ioLock = new Object();
+        private volatile IOControl savedIoControl;
+
+        private ApacheClientBackpressure(HttpSettings settings) {
+            this.settings = settings;
+        }
+
+        private void addBytesAndMaybePause(long count, IOControl ioControl) {
+            if (bytesInQueue.accumulateAndGet(count, Long::sum) >= settings.getMaxResponseSize().getBytes()) {
+                pauseProducer(ioControl);
```

Comment:
Is it possible that the `addBytesAndMaybePause` could be called again after the queue is already full (aka such that the if-block would return true)? Would that matter? I assume the most recent `IOControl` supersedes any that we've set previously?